ENH: Implement string comparison ufuncs (or almost)

seberg opened the pull request:
This makes all comparison operators and ufuncs work on strings
using the ufunc machinery.
It requires a half-manual "ufunc" to keep supporting void comparisons
and especially `np.compare_chararrays` (that one may have a bit more
overhead now).

In general the new code should be much faster, and has a lot of easier
optimization potential.  It is also much simpler since it can outsource
some complexities to the ufunc/iterator machinery.

This further fixes a couple of bugs with byte-swapped strings.

The backward compatibility related change is that using the normal
ufunc machinery means that string comparisons between string and
unicode now give a `FutureWarning` (instead of just False).

---

This also shows two things:
1. How to use new-style loops (very simple one)
2. Writes them in C++ templating.

I do admit, that I think we will have to refactor some of this when we get more ufuncs created in this way.  Note that for now `np.equal.types` will *not* list `SS->?` and `UU->?`.

In general, this is finished, however, it should get a bit of new tests for direct `np.equal`, etc. ufunc use and for byte-swapped or unaligned unicode inputs.

seberg committed:
ENH: Implement string comparison ufuncs (or almost)

This makes all comparison operators and ufuncs work on strings
using the ufunc machinery.
It requires a half-manual "ufunc" to keep supporting void comparisons
and especially `np.compare_chararrays` (that one may have a bit more
overhead now).

In general the new code should be much faster, and has a lot of easier
optimization potential.  It is also much simpler since it can outsource
some complexities to the ufunc/iterator machinery.

This further fixes a couple of bugs with byte-swapped strings.

The backward compatibility related change is that using the normal
ufunc machinery means that string comparisons between string and
unicode now give a `FutureWarning` (instead of just False).

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,450 @@
+    NPY_NO_EXPORT PyObject *
+    _umath_strings_richcompare(
+            PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip);
+}
```
Is this how we do this for C++?  I am not sure which header I could cram this in so that i can just include it from `multiarray/multiarrayobject.c` and `multiarray/arrayobject.c` which need access to these.  Currently I put it there directly, which is OK, but I don't like it too much.

serge-sans-paille replied:
Why not in an `string_ufuncs.h` ?

seberg commented:
@serge-sans-paille could you vet the C++ again?  Maybe you also know how to reduce those annoying repetitions when creating the many different loops?  That would be awesome :).

seberg commented on numpy/core/src/multiarray/convert_datatype.h:
```diff
@@ -62,7 +66,7 @@ NPY_NO_EXPORT int
 
 NPY_NO_EXPORT int
-PyArray_AddCastingImplementation_FromSpec(PyArrayMethod_Spec *spec, int private);
+PyArray_AddCastingImplementation_FromSpec(PyArrayMethod_Spec *spec, int private_);
```
C++ doesn't like "private"... even in `extern "C"` apparently?

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,451 @@
+            NPY_ITER_WRITEONLY | NPY_ITER_ALLOCATE | NPY_ITER_ALIGNED};
+
+    PyArrayMethod_StridedLoop *strided_loop = nullptr;
+    NPY_BEGIN_THREADS_DEF;
```
C++ forced me to move all the inits up, or it didn't want the goto, another C99 feature not included in C11?

serge-sans-paille replied:
I wonder if flagging these variables as `constexpr` changes something wrt. `goto` ?

seberg replied:
Just tried for the flags, and it didn't seem to work.

mattip commented:
Windows is [not happy](https://dev.azure.com/numpy/numpy/_build/results?buildId=22828&view=logs&j=c499766f-449b-5541-ed00-14b3885196ca&t=c16bbb4c-8d4f-5cec-143b-0575175479e7&l=1081).

Do we have benchmarks that are affected by this change?

seberg committed:
MAINT: Do not use C99 tagged struct init in C++

C++ does not like it (at least not before C++20)... GCC and clang
don't seem to mind, but MSVC seems to.

mhvk commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+{
+    /*
+     * Note, this works in CPython even without the GIL, however it may be that
+     * this will have to be moved into `auxdata` eventually, which may be
```
What does "this" refer to? The calculations of `len1` and `len2`? Here, too, I'd try to make the comment more like a TODO or possible one "TODO?" -- as is, the comment seems to presume a question about why something is not done some other way, but I don't know what the question is!

seberg replied:
Oh, it refers to fetching the `elsize`.

mhvk commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+ * and registers it with the given ufunc.
+ */
+static int
+add_loop(
```
Total nitpick, but why the new line?

seberg replied:
ah, because right now I tend to 8-space hanging indent, but that feels strange if there is more than one "special" parameter on the other line (and it is 1 space off here).  Anyway, just removed it, and aligned the second line with the bracket here...

mhvk commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,444 @@
+}
+
+NPY_NO_EXPORT PyObject *
+_umath_strings_richcompare(
```
Just so I understand, this function exists only because of the `rstrip` that is needed for `chararray`, is that correct? And the regular string array comparison will just go through the ufunc. If so, good to add a comment to make that explicit.

In the new scheme of things, would one define a new string dtype and then add loops for it that do auto-rstrip? (Not that I think that is a good idea -- seems more sensible to just rstrip on input!).

mhvk replied:
I agree it is not worth keeping - just wondered what in principle the method would be to support it. And of course good to mention the `void` comparisons in a comment!

mhvk reviewed (commented):
@seberg - I had a look mostly just because I was curious both how your new dtypes held up here and how the new templating works. Both seem great!  My comments are really just about trying to make comments clearer; you probably want one of the more c++ people to have a look.

seberg committed:
BENCH: Add basic string comparison benchmarks

seberg committed:
DOC,STY: Fixup string-comparisons comments based on review

Thanks to Marten's comments, a few clarfications and slight fixups.

seberg committed:
ENH: Use `memcmp` because it may be faster for the byte case

seberg commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -83,6 +83,10 @@ NPY_NO_EXPORT int NPY_NUMUSERTYPES = 0;
+/* From umath/string_ufuncs.cpp/h */
+NPY_NO_EXPORT PyObject *
+_umath_strings_richcompare(
+        PyArrayObject *self, PyArrayObject *other, int cmp_op, int rstrip);
```
There be dragons...  I am not quite sure how big they are but there are difficulties with crossing the multiarray/umath boundary.

seberg commented on numpy/core/src/multiarray/convert_datatype.h:
```diff
@@ -34,14 +38,14 @@ dtype_kind_to_ordering(char kind);
 NPY_NO_EXPORT npy_bool
 can_cast_scalar_to(PyArray_Descr *scal_type, char *scal_data,
-                    PyArray_Descr *to, NPY_CASTING casting);
+        PyArray_Descr *to, NPY_CASTING casting);
```
oh, should just undo this, there was an accidental auto-reformat due to adding the `extern "C" {`.

seberg commented on numpy/core/include/numpy/experimental_dtype_api.h:
```diff
@@ -214,7 +214,7 @@ typedef struct {
 
 
-typedef PyObject *_ufunc_addloop_fromspec_func(
+typedef int _ufunc_addloop_fromspec_func(
```
True, although I don't care about when/how it gets in, hehe

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,444 @@
+template <typename character>
+static NPY_INLINE int
+character_cmp(character a, character b)
+{
```
Yeah, although I am not sure in case of invalid unicode characters (they are probably unsigned, but that would be out of range).  Happy to change if think it makes it much clearer?  (I don't have an opinion either way.)

serge-sans-paille replied:
yeah, maybe something other than an in is better as `typename std::intermediate_type<npy_ucs4, npy_byte>::type`

seberg replied:
I honestly still don't really see that this will lead to easier to read code :), and I assume the compilers should optimize it just fine?

serge-sans-paille replied:
Not quite, and that's why I'm trying to convince you :-)

https://godbolt.org/z/GMWPKE4EK

seberg replied:
Yeah, but that is not what we are comparing?  We are comparing these:
https://godbolt.org/z/f5qh9a3b7

And there `func1` actually ends up shorter in some cases (depends on the operand)?

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,444 @@
+    while (N--) {
+        int cmp = string_cmp<rstrip>(
+                len1, (character *)in1, len2, (character *)in2);
+        npy_bool res;
```
Sure, lets go with a switch then (will followup soon).

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,444 @@
+        else if (comp == Py_GE) {
+            res = cmp >= 0;
+        }
+        else {
```
Yeah, I thought so first, but my linter didn't like `static_assert(0)`?  But probably just a bad linting hint, not fully understanding the templating, will change.

seberg replied:
That is what I thought first, but I just tried again and both my linter (and now I saw even the compiler) did not like a `static_assert(false)`?

seberg committed:
TST: Improve string and unicode comparison tests.

seberg committed:
MAINT: Use switch statement based on review

As suggested be Serge.

Co-authored-by: Serge Guelton <serge.guelton@telecom-bretagne.eu>

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,444 @@
+    PyArrayMethod_StridedLoop *loop;
+
+    /* TODO: It would be nice to condense the below */
+    /* All String loops */
```
:thinking: I have to admit, that is utterly non-obvious to me right now, I seem to be missing some concepts to make easy sense of the example (but I admit, I did not take more than 10 minutes to try to understand it).

seberg replied:
Ah, now I get it.  But to do this well, I would need to pass "tuples" with `"equal", Py_EQ`, I guess.  At least unless I double down on the "equal" spelling and use that everywhere rather than the int.

serge-sans-paille replied:
The function `int_to_string` would create the binding between `comp` and the string name. If you prefer, I'll propose a patch once this one lands and I'll add you... as a reviewer ;-)

seberg replied:
Hmm, I half see it, but it would be awesome if we could do something like a clear mapping somewhere, maybe not spelled as a switch statement? If you are up to a followup, that is great :).

seberg commented:
IMO, the main point here is making the string comparisons sane and adding them as ufuncs to `np.equal`, etc. to begin with.  But benchmarks showed a slightly different performance "profile" for `S` with more characters and when things match.  I am not convinced that it is actually _better_ how it was before (unless we anticipate long strings), but for now changed it, more than happy to revert though!

<details> <summary> <b> The benchmarks proof my point that this is "obviously" much faster: </b> </summary>

(This time, smaller ratio is better)
```
       before           after         ratio
     [d9538400]       [623cea57]
     <main>           <string-ufuncs>
+         104±1μs        118±0.4μs     1.14  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '!=')
+      54.6±0.9μs      60.3±0.09μs     1.10  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '!=')
+         223±7μs          237±2μs     1.06  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '!=')
-       131±0.3μs        122±0.7μs     0.94  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '!=')
-     1.27±0.01μs      1.18±0.02μs     0.93  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '!=')
-     1.79±0.01μs      1.40±0.01μs     0.78  bench_strings.StringComparisons.time_compare_identical(100, 'U', True, '<=')
-     1.30±0.03μs      1.01±0.01μs     0.78  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '<=')
-     1.29±0.02μs         999±10ns     0.78  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '==')
-      53.8±0.5μs       41.7±0.3μs     0.78  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '>=')
-     1.80±0.01μs      1.40±0.01μs     0.77  bench_strings.StringComparisons.time_compare_identical(100, 'U', True, '>=')
-        1.28±0μs         991±20ns     0.77  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '>')
-     1.78±0.01μs      1.38±0.01μs     0.77  bench_strings.StringComparisons.time_compare_identical(100, 'U', True, '>')
-     1.28±0.01μs          984±4ns     0.77  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '>=')
-        1.27±0μs          975±2ns     0.77  bench_strings.StringComparisons.time_compare_identical(100, 'U', False, '<')
-        54.3±2μs       41.6±0.3μs     0.77  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '<=')
-         970±8ns          740±3ns     0.76  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '>')
-       213±0.4μs          162±2μs     0.76  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '>=')
-        54.1±2μs       41.1±0.3μs     0.76  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '>')
-         104±1μs       79.0±0.3μs     0.76  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '<=')
-     1.22±0.01μs          926±4ns     0.76  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '>=')
-      52.9±0.5μs       40.1±0.5μs     0.76  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '==')
-         976±5ns         739±20ns     0.76  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '<=')
-     54.2±0.06μs       40.7±0.1μs     0.75  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '<')
-        1.81±0μs      1.36±0.01μs     0.75  bench_strings.StringComparisons.time_compare_identical(100, 'U', True, '==')
-         106±6μs       79.5±0.6μs     0.75  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '>=')
-        984±20ns          738±5ns     0.75  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '>=')
-       106±0.5μs       79.2±0.4μs     0.75  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '>')
-         104±1μs       77.7±0.6μs     0.75  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '==')
-        966±10ns          719±4ns     0.74  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '==')
-     1.85±0.02μs      1.36±0.02μs     0.74  bench_strings.StringComparisons.time_compare_identical(100, 'U', True, '<')
-       966±0.7ns          711±2ns     0.74  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '<=')
-        1.23±0μs          901±6ns     0.73  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '>')
-        982±10ns          721±3ns     0.73  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '>')
-         108±1μs      78.8±0.06μs     0.73  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '<')
-     1.19±0.03μs          873±8ns     0.73  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '<')
-     1.22±0.01μs          894±6ns     0.73  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '<=')
-         974±7ns          713±2ns     0.73  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '<')
-     1.20±0.02μs          874±6ns     0.73  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '!=')
-         970±1ns          706±6ns     0.73  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '!=')
-         970±9ns          705±5ns     0.73  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '==')
-     1.22±0.02μs          887±4ns     0.72  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '>=')
-     1.21±0.01μs          878±8ns     0.72  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '<=')
-         977±8ns          707±4ns     0.72  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '<')
-     1.22±0.01μs          884±2ns     0.72  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '>')
-         218±7μs          158±2μs     0.72  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '<=')
-         218±7μs        158±0.9μs     0.72  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '<')
-     1.23±0.01μs          890±4ns     0.72  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '<')
-        996±30ns          718±2ns     0.72  bench_strings.StringComparisons.time_compare_different(100, 'S', False, '>=')
-         978±5ns          698±2ns     0.71  bench_strings.StringComparisons.time_compare_identical(100, 'S', False, '!=')
-     1.25±0.02μs          891±4ns     0.71  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '==')
-         228±1μs          160±1μs     0.70  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '>')
-         226±1μs          158±1μs     0.70  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', True, '==')
-     1.23±0.03μs          859±5ns     0.70  bench_strings.StringComparisons.time_compare_identical(100, 'S', True, '!=')
-     1.24±0.04μs         850±10ns     0.68  bench_strings.StringComparisons.time_compare_different(100, 'S', True, '==')
-         115±7μs       78.1±0.2μs     0.68  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '>=')
-      59.4±0.8μs      39.5±0.09μs     0.66  bench_strings.StringComparisons.time_compare_identical(10000, 'S', True, '>=')
-      31.2±0.1μs       20.1±0.6μs     0.65  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '>=')
-         113±6μs       72.5±0.3μs     0.64  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '==')
-       128±0.5μs         81.6±1μs     0.64  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '>=')
-       130±0.8μs       82.6±0.9μs     0.64  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '==')
-        53.9±3μs       34.0±0.2μs     0.63  bench_strings.StringComparisons.time_compare_different(10000, 'S', True, '<')
-       131±0.5μs       82.4±0.5μs     0.63  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '<=')
-        55.1±5μs       34.6±0.3μs     0.63  bench_strings.StringComparisons.time_compare_different(10000, 'S', True, '<=')
-        58.4±1μs       36.5±0.1μs     0.63  bench_strings.StringComparisons.time_compare_identical(10000, 'S', True, '<=')
-         966±7ns          603±3ns     0.62  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '<')
-       131±0.9μs       81.6±0.8μs     0.62  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '>')
-        29.4±2μs       18.3±0.6μs     0.62  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '>')
-        972±20ns          603±5ns     0.62  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '!=')
-         131±2μs       81.1±0.6μs     0.62  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'U', False, '<')
-        59.3±2μs       36.6±0.3μs     0.62  bench_strings.StringComparisons.time_compare_identical(10000, 'S', True, '==')
-         954±6ns          586±2ns     0.61  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '<=')
-        971±40ns          587±4ns     0.60  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '>=')
-        977±10ns          590±1ns     0.60  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '==')
-      27.3±0.7μs       16.5±0.2μs     0.60  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '<')
-        30.3±2μs       18.3±0.6μs     0.60  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '<=')
-         113±7μs       67.8±0.5μs     0.60  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '>')
-         988±4ns          593±1ns     0.60  bench_strings.StringComparisons.time_compare_different(100, 'U', False, '>')
-        51.9±2μs         31.1±1μs     0.60  bench_strings.StringComparisons.time_compare_different(10000, 'S', True, '==')
-     31.2±0.09μs       18.6±0.5μs     0.60  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '==')
-        59.6±4μs         35.4±2μs     0.59  bench_strings.StringComparisons.time_compare_identical(10000, 'S', True, '>')
-       122±0.2μs       72.3±0.2μs     0.59  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '>')
-         116±6μs       68.2±0.2μs     0.59  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '!=')
-      31.3±0.5μs       18.3±0.5μs     0.59  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '<=')
-       122±0.6μs         71.1±3μs     0.58  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '<=')
-      30.2±0.6μs       17.5±0.9μs     0.58  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '<')
-        70.7±2μs      40.7±0.07μs     0.58  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '>=')
-        1.21±0μs         697±20ns     0.57  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '!=')
-         112±7μs         64.3±3μs     0.57  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '>=')
-        59.5±1μs       33.8±0.2μs     0.57  bench_strings.StringComparisons.time_compare_identical(10000, 'S', True, '!=')
-         114±9μs         64.7±2μs     0.57  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '!=')
-        30.3±1μs       17.1±0.5μs     0.57  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '!=')
-      31.0±0.7μs      17.5±0.08μs     0.56  bench_strings.StringComparisons.time_compare_identical(10000, 'S', False, '!=')
-         119±6μs         67.0±3μs     0.56  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', True, '<')
-        60.4±2μs       33.8±0.2μs     0.56  bench_strings.StringComparisons.time_compare_different(10000, 'S', True, '!=')
-        60.8±3μs       34.0±0.4μs     0.56  bench_strings.StringComparisons.time_compare_different(10000, 'S', True, '>=')
-         118±3μs         65.1±2μs     0.55  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '<')
-        30.9±2μs       17.0±0.7μs     0.55  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '>')
-      31.1±0.2μs       17.0±0.6μs     0.55  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '>=')
-     1.22±0.01μs          667±2ns     0.55  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '<')
-        1.23±0μs          668±5ns     0.54  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '>=')
-     1.23±0.01μs          666±7ns     0.54  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '==')
-     1.25±0.02μs         673±10ns     0.54  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '<=')
-         120±3μs         64.6±3μs     0.54  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '<=')
-     1.24±0.02μs          665±1ns     0.54  bench_strings.StringComparisons.time_compare_different(100, 'U', True, '>')
-        29.6±2μs       15.8±0.5μs     0.53  bench_strings.StringComparisons.time_compare_different(10000, 'S', False, '==')
-     25.3±0.08μs      13.1±0.07μs     0.52  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '>=')
-        74.1±3μs      38.0±0.07μs     0.51  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '==')
-      25.2±0.3μs       12.9±0.1μs     0.51  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '>')
-      25.3±0.2μs      12.9±0.08μs     0.51  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '<=')
-     25.0±0.08μs       12.7±0.1μs     0.51  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '<')
-         117±3μs         59.1±3μs     0.50  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', True, '==')
-      25.0±0.1μs       12.5±0.2μs     0.50  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '==')
-        75.7±3μs         37.4±1μs     0.49  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '>')
-      25.6±0.6μs       12.6±0.1μs     0.49  bench_strings.StringComparisons.time_compare_different(10000, 'U', False, '!=')
-        72.7±2μs         34.2±1μs     0.47  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '>=')
-      75.8±0.4μs      35.3±0.05μs     0.47  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '<=')
-        75.7±2μs       35.1±0.1μs     0.46  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '<=')
-        76.0±2μs       34.9±0.1μs     0.46  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '<')
-        76.4±2μs       35.1±0.2μs     0.46  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '<')
-        76.0±3μs         34.0±1μs     0.45  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '!=')
-        78.6±1μs         33.9±1μs     0.43  bench_strings.StringComparisons.time_compare_identical((1000, 20), 'S', False, '!=')
-        78.8±3μs         33.9±1μs     0.43  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '>')
-      60.4±0.2μs       25.7±0.1μs     0.43  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '>=')
-        76.1±2μs         31.3±1μs     0.41  bench_strings.StringComparisons.time_compare_different((1000, 20), 'S', False, '==')
-      62.5±0.4μs       25.7±0.1μs     0.41  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '<=')
-      63.0±0.3μs       25.7±0.1μs     0.41  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '>')
-      62.5±0.7μs      25.1±0.03μs     0.40  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '==')
-        64.6±2μs      25.6±0.07μs     0.40  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '<')
-        64.8±2μs       25.5±0.2μs     0.39  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', False, '!=')
-      48.7±0.3μs       18.8±0.4μs     0.39  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '<=')
-      48.9±0.2μs       18.6±0.2μs     0.38  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '<')
-      48.6±0.2μs       18.4±0.1μs     0.38  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '!=')
-      49.0±0.6μs       18.5±0.2μs     0.38  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '==')
-      48.9±0.5μs      18.4±0.09μs     0.38  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '>')
-      49.4±0.6μs       18.5±0.2μs     0.38  bench_strings.StringComparisons.time_compare_different(10000, 'U', True, '>=')
-      97.0±0.3μs       36.2±0.2μs     0.37  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '!=')
-      96.9±0.7μs       35.9±0.2μs     0.37  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '<')
-      98.0±0.4μs       36.2±0.8μs     0.37  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '<=')
-      97.9±0.4μs       36.1±0.4μs     0.37  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '>=')
-      97.9±0.3μs       35.7±0.7μs     0.36  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '>')
-      98.1±0.8μs       35.6±0.3μs     0.36  bench_strings.StringComparisons.time_compare_different((1000, 20), 'U', True, '==')
```

</details>

seberg committed:
TST: Make unicode byte-swap test slightly more concrete

The issue is that the `view` needs to use native byte-order, so
just ensure native byte-order for the view, and then do another cast
to get it right.

mhvk commented:
@seberg - those are very impressive improvements!! I'm rather puzzled by this, though:
```
+         104±1μs        118±0.4μs     1.14  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '!=')
-         104±1μs       77.7±0.6μs     0.75  bench_strings.StringComparisons.time_compare_identical(10000, 'U', True, '==')
```
How is it possible that '!=' is 50% slower than '=='?? I would have expected the two to be identical (as was the case in the old code). And indeed, all the other comparisons are the same (within uncertainties) as '=='.

The same 50% penalty holds when not using `rstrip`, so that is not the problem:
```
+      54.6±0.9μs      60.3±0.09μs     1.10  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '!=')
-      52.9±0.5μs       40.1±0.5μs     0.76  bench_strings.StringComparisons.time_compare_identical(10000, 'U', False, '==')
```
As well as for the 2-d case.

seberg commented:
Hmmm, no idea, without thinking about it, I would have assumed that all comparisons will have the same speed.  Is there some reason the compiler can optimize some path more than the other, or just random?

mhvk commented:
The timing differences are very systematic: all comparisons are the same *except* '!=' for 'U'. I cannot see any reason for it in your code; after all, the base function doing the comparisons does not even know what comparison is being done; it just returns -1, 0, or +1. Maybe it is some kind of weird fluke?

mhvk commented:
I.e., maybe worth running the benchmarks again?

seberg commented:
Its reproducible on my comp, also if I change the order in which the benchmarks run.  The compiler must be spitting something different for `!=`.

mhvk commented:
Weird, I just thought I'd try to reproduce it too, and I cannot:
```
In [1]: import numpy as np

In [2]: a = np.arange(100000.0).astype("U")

In [3]: b = a.copy()

In [4]: %timeit a != b
2.23 ms ± 10.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: %timeit a == b
2.29 ms ± 11.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
(For both, current master gives 3.46 ms, so I confirm the great speed-up! I also tried with `np.arange(10000)`, just to be closer to the benchmark.)

seberg commented:
To be honest, I currently don't care much about this oddness related to the core of the loop right now.  Could be compiler version or even hardware differences (gcc 11.2 and a Ryzen 3 here) :shrug:.
But if it is specific to my hardware, maybe that adds a reason to not worry about it?

If anyone to squeeze performance, I would suggest implementing the `get_loop` and looking into specializing it for short/long strings (and even contiguous+short or so), I also didn't even try adding `NPY_GCC_OPT3`.  Is that worthwhile for string arrays?

mhvk commented:
@seberg - agreed that in the larger scheme of things this just doesn't matter. The code is much nicer and faster, what else could one ask for!?

seberg committed:
BUG: Add `np.compare_chararrays` to test and fix typo

seberg committed:
TST: Add test for empty string comparisons

mhvk commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,457 @@
+}
+
+
+extern "C" {
```
Like below: I think it makes sense to create a 'strings_ufunc.h' that contains these definitions.

seberg replied:
Maybe @mattip, you have a good idea?  I had that first, and it should work for the `init_string_ufuncs`, but the umath/multiarray boundary is always creating some headaches: it seems hard to include that header from `multiarray` somehow?

I admit, it is very likely I am just doing something stupid.

Also @serge-sans-paille, I thought another problem was that the symbol export didn't actually work right when I had it as its own header here.  But I guess that may have been just a bad try?

mhvk replied:
Maybe good to separate what should more obviously work, which is to put the ones double-defined here in 'strings_ufunc.h' and import that new .h above. And only then worry about the other places. 

Regarding multiarray/umath, I thought we had tried to harmonize those, but not sure...

seberg replied:
Yes, they are now one build-unit, which is why this works at all without much hassle.  But including umath stuff in multiarray is still tricky (I am not exactly sure why)

seberg replied:
OK, moved things back, it is working fine, but did not try to cross the `multiarray` boundary yet.  I guess one thing could be that I did not actually include other stuff in the header right now (not sure if I should).

mhvk commented on numpy/core/src/umath/umathmodule.c:
```diff
@@ -24,6 +24,10 @@
 #include "number.h"
 #include "dispatching.h"
 
+/* From string_ufuncs.cpp */
```
If one defines 'strings_ufuncs.h', it could just be included here.

mhvk commented on numpy/core/tests/test_strings.py:
```diff
@@ -0,0 +1,85 @@
+    res1 = ufunc(arr_string, arr_unicode, signature="UU->?", casting="unsafe")
+    res2 = ufunc(arr_string, arr_unicode, signature="SS->?", casting="unsafe")
+
+    expected = [op("a", "a"), op("a", "c")]
```
Hah, the dangers of hard-coding! Second one should be `op("b", "c")`. Or how about `expected = op(arr_string.astype('U'), arr_unicode)`?

mhvk reviewed (commented):
Looks good! Mostly just a reiteration of @serge-sans-paille's suggestion to create an 'string_ufuncs.h', plus a small nit on one of the tests.

Though I'd prefer a more C++ adept person to sign off on that part!

seberg committed:
TST: Fixup string test based on martens review

seberg committed:
MAINT: Move definitions back into string_ufuncs.h

serge-sans-paille commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+                res = cmp >= 0;
+                break;
+            default:
+                assert(false);
```
the right `static_assert` here would be (in the function prelude`

```
static_assert(Py_EQ == comp || Py_NE == comp || ..., "valid comparison");
```
or something based on a range if you know for sure which enumerator is the lowest one.

serge-sans-paille commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+        case Py_GE:
+            return string_comparison_loop<rstrip, Py_GE, character>;
+        default:
+            assert(false);
```
ok, seeing the same assert again, what you probably should do is

```
static constexpr is_valid_comp(int comp) {
   return comp == Py_LT || ...;
}
```
and use it either in fonction prelude, either as an `assert` or a `static_assert` depending on the context

Or even better: make the int an `enum class` but that's probably not possible, just mentioning it.

seberg replied:
OK, I didn't love the idea of a long duplicate `Py_EQ || ...` where the asserts were, but if it goes into the function prelude, I can see the point, will adopt :).

seberg replied:
clang static analysis seems to not be quite smart enough to realize that the switch will always match, but I guess I will ignore that (it warns that `res` may be undefined in the other case).

seberg replied:
I guess I could use `default` for the last one, but dunno.

seberg replied:
Oh, silly me... This assert is actually not static at all.

seberg replied:
Sorry for the verboseness. Pushed a change with that the enum class instead, figured that the static assert was just as verbose anyway...

seberg committed:
MAINT: Use enum class for comparison operator templating

This removes the need for a dynamic (or static) assert in the
switch statement.

serge-sans-paille committed:
Template version of add_loop to avoid redundant code

serge-sans-paille commented:
@seberg : feel free to cherry-pick https://github.com/serge-sans-paille/numpy/commit/0bf1a47e3236dded66b992565ce742d8a1048f49

seberg committed:
STY: Fixup style, two spaces, error is -1

seberg commented:
Done that cherry pick, but also an STY commit because some style stuff I just can't agree with ;).  Left some long lines, wasn't sure how to break them, and I think they are somewhat fine (i.e. nothing non-obvious hidden), although not great.

I am happy with this, it is nice that it removes the duplication, it adds a bit of "C++" knowledge requirement to truly understand what is going on, but I guess that is OK.

seberg commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+        case COMP::GE: return "greater_equal";
+    }
+    assert(0);
+    return nullptr;
```
Had to add this to silence GCC warning, clang tells me its unreachable (correct) gcc warns it may return undefined stuff (I guess in principle also correct if a bad `comp` value is passed at runtime).

serge-sans-paille replied:
more of a styling issue, but could that be but in a `default:` from the switch instead?
In case the Numpy dev would be interested, there's a ``__builtin_unreachable()`` builtin to cope with this kind of situation, e.g.

```
/// LLVM_BUILTIN_UNREACHABLE - On compilers which support it, expands
/// to an expression which states that it is undefined behavior for the
/// compiler to reach this point.  Otherwise is not defined.
///
/// '#else' is intentionally left out so that other macro logic (e.g.,
/// LLVM_ASSUME_ALIGNED and llvm_unreachable()) can detect whether
/// LLVM_BUILTIN_UNREACHABLE has a definition.
#if __has_builtin(__builtin_unreachable) || defined(__GNUC__)
# define LLVM_BUILTIN_UNREACHABLE __builtin_unreachable()
#elif defined(_MSC_VER)
# define LLVM_BUILTIN_UNREACHABLE __assume(false)
#endif
```

seberg replied:
Yeah, moving it into the `default` clause worked.  I am tempted to leave the idea of using `__builtin_unreachable` to later.  I don't mind using it, but we should probably introduce it to a few places in one go.
Although in some places having an `assert` may also be nice in debug mode (not sure if `builtin_unreachable` does that in debug mode).

seberg commented:
Just stumbled on this again.  I realize that the comparison code may still have to small optimization with casting available.

But should we just move ahead with this, or does anyone recall any downsides.  There was this one note:

> The backward compatibility related change is that using the normal ufunc machinery means that string comparisons between string and unicode now give a FutureWarning (instead of just False).

mhvk commented:
I was OK with it already, but maybe @serge-sans-paille can have a last look?

serge-sans-paille commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+ */
+template <int rstrip, typename character>
+static NPY_INLINE int
+string_cmp(int len1, character *str1, int len2, character *str2)
```
I don't know what's the convention wrt. const-correctness in Numpy, but `str1` and `str2` could be pointers to `const character`

serge-sans-paille commented on numpy/core/src/umath/string_ufuncs.cpp:
```diff
@@ -0,0 +1,448 @@
+ * zero padded (trailing zeros are ignored in other words, the shorter word
+ * is always padded with zeros).
+ */
+template <int rstrip, typename character>
```
`rstrip` is a  `bool` in other spots.

serge-sans-paille reviewed (commented):
No blocking comments, just a few nits.

seberg committed:
STY: Small `string_ufuncs.cpp` fixups based on Serge's review

seberg committed:
MAINT: Fix merge conflict (ensure_dtype_nbo was removed)

seberg commented:
@charris just a gentle ping, since I think this was pretty much ready to go in and get tested.

dlee992 commented:
@seberg, hi. Recently I want to contribute a PR about string ufunc in Numba (https://github.com/numba/numba/pull/8104), which is empty PR right now. I discussed with Numba members, and we hope to keep consistent with NumPy handling way. And I saw this PR! Very impressive and we need it (especially `SS->?` and `UU->?` and unify string with other types in ufunc loop)! 
I wonder whether this PR can be merged into NumPy 1.23.0 version ASAP, since I am pending on this very much. Haha. Thanks!

charris commented:
Yeah, let's give this a shot. Thanks Sebastian. I'm going to do an rc3 this weekend.

Looks like this needs a release note.  @seberg Since I will backport this for @dlee992, could you add the release note directly against the 1.23.x branch after this is backported?

charris commented:
The backport is in #21716.

seberg commented:
Oh we are giving a backport a shot?  Will there be another (brief) rc?  I don't really expect this to create issues, but it might have some very subtle changes.

I don't mind backporting, I do wonder though. @dlee992 are you interested in NumPy *supporting* `np.equal()`, etc. on the strings.  Or do you need the API to extend NumPy ufuncs yourself in that way.  Because the API here is only public experimentally right now.

charris commented:
> Will there be another (brief) rc?

Yes, there are already > 10 backports since rc2. And then I just want to sit on it for two weeks or so and stop doing that :) The zero release is the true NumPy rc, since that is when widespread testing really happens.

dlee992 commented:
> I don't mind backporting, I do wonder though. @dlee992 are you interested in NumPy _supporting_ `np.equal()`, etc. on the strings. Or do you need the API to extend NumPy ufuncs yourself in that way. Because the API here is only public experimentally right now.

@seberg, hi. I think I just need NumPy _supporting_ `np.equal()`, etc. on the strings, not to extend other NumPy ufuncs by myself. BTW, I'd guess there will be other PRs to add `SS->?`/`UU->?` into `np.equal.types` etc? Since numba also relies on `np.equal.types` to implement its ufunc capabilities (maybe because it was implemented by Pitrou several years ago, which I am not sure).

> If @dlee992 decides that a backport will not serve, the we can close the open PR for that.

@charris, hi, I need the very backport. I am glad it can be backported to 1.23.0.

Thanks for ur quick response very much!

seberg commented:
@dlee992 `np.equal.types` is unmodified right now, I had not planned on finding a solution to that at this time.  I suppose we could try (there may not be a _complete_ solution, so I expect we may need to add new Python/C-API for numba in any case if you need this).

We should be able to expand the output of `np.equal.types` (but not the binary version stored on the ufunc struct) in some of the clear cases (such as `np.equal.types`).

Note that `np.equal.types` lists old-style loops, but these are new-style loops (they have a different signature).

dlee992 commented:
> @dlee992 `np.equal.types` is unmodified right now, I had not planned on finding a solution to that at this time. I suppose we could try (there may not be a _complete_ solution, so I expect we may need to add new Python/C-API for numba in any case if you need this).

@seberg, maybe I can try to follow these parts of NumPy code, and try to extend `np.equal.types` for NumPy. But I am a newbie for NumPy, I will give it a try.

> We should be able to expand the output of `np.equal.types` (but not the binary version stored on the ufunc struct) in some of the clear cases (such as `np.equal.types`).

I'd guess these related stuff are implemented in `string_ufuncs.cpp` and `arrayobject.c`? IIRC, numba only uses `np.equal.types` as a string list to select an appropriate signature, and nothing more (not quite sure), it will generate binary compiled version or sth by itself (some discussion is here numba/numba#7859).

seberg commented:
> @seberg, maybe I can try to follow these parts of NumPy code, and try to extend np.equal.types for NumPy

Of course, the problem is timing.  If we really need this in the next release we have to do it super-fast and convince Chuck that backporting matters.  It is probably not hard, but it seems a bit steep to me.
There may also be some subtleties, although for you it just matters to have the strings there (i.e. not a full solution, but something good enough for you).

> I'd guess these related stuff are implemented in string_ufuncs.cpp and arrayobject.c? IIRC, numba only uses np.equal.types as a string list to select an appropriate signature, and nothing more (not quite sure)

This happens in `umath/ufunc_object.` there is a getter for the types.  The problem is that the string loops have a different ABI then the "legacy" loops.  And the `ufunc.types` are a summary of the "legacy" loops and do not contain the new ABI loops.
Even worse, a list of strings is probably not sufficient for a future-proof exposing in any case (i.e. it may work for strings, but not generally).

A "workaround" might be to add a "legacy" loop that is never used and just does nothing/sets an error if used.  But that is pretty annoying.
The other workaround is to just manually extend the Python `ufunc.types` to include some of the new loops (if possible to describe them using simple strings).

dlee992 commented:
@seberg, many thanks! But in fact, I cannot follow ur thought about extending `ufunc.types` right now, I have read Numba doc several times and open four PRs for Numba, but I just began to read NumPy doc e.g., [Numpy C-API doc](https://numpy.org/doc/stable/reference/c-api/index.html), and build NumPy from source. I have to get a big picture about various concepts used in NumPy more carefully, then I guess I can follow this special `string ufunc` stuff and your workaround thought. I will spend several days in these parts and be back after this process. Thanks again!

PS, I think I am not that hurry about this feature in NumPy 1.23.0 (maybe 1.23.1 is also OK for me), we can do it better with more thoughtful discussion.

seberg commented:
If there is consensus that returning two extra entries in `np.equal.types` we could do that later maybe, the question is whether we should worry about any backwards compatibility issue (say because someone auto-generates things from it, similar to numba).
In that case, I am not sure if it is a "bugfix" to land in 1.23.1 as opposed to 1.24  (we have a half-year release cadence).

We are now in the third rc for 1.23, so I would not expect many more changes unless something comes up.  In the end it is up to Chuck though.

The following is what returns `ufunc.types` to Python:
https://github.com/numpy/numpy/blob/703a6fa6ac7f5049bb17726a1ba93f7835bf92ae/numpy/core/src/umath/ufunc_object.c#L6336-L6368

As you can see it reads `ufunc->types`.  But NumPy now also has another list of "loops", which is stored as an actual list in `ufunc->_loops`.  Now, we could scan/use `ufunc->_loops` rather than `ufunc->types`.  The problems are:
* The `ufunc->_loops` can have additional entries we need to skip.
* In principle at least, `ufunc->_loops` could have entries that are *not* representable by single characters (breaking the formatting).
* `ufunc->_loops` can get outdated for numba's DUFuncs, so you probably still need to have a check on `ufunc->types` (at least for changes).

seberg commented:
@dlee992 for your awareness.  It seems we should (and will) pull this out of NumPy 1.23.  It leads to subtle changes for subclasses (subclasses were sometimes ignored before).  That leads to subtle changes/breaks in MaskedArrays paths (that were arguably always a bit bad).  And those breaks domino into failures for `astropy`...
